File: app/src/main/java/com/shevelev/comics_viewer/common/threads/CancelationToken.java
package com.shevelev.comics_viewer.common.threads;

import java.util.concurrent.locks.ReentrantLock;

public class CancelationToken implements ICancelationTokenRead
{
    private boolean isCanceled=false;
    private final ReentrantLock lock = new ReentrantLock();

    @Override
    public boolean isCanceled()
    {
        boolean isCanceled=false;

        lock.lock();
        try
        {
            isCanceled=this.isCanceled;
        }
        finally
        {
            lock.unlock();
        }

        return isCanceled;
    }

    public void cancel()
    {
        lock.lock();
        try
        {
            this.isCanceled=true;
        }
        finally
        {
            lock.unlock();
        }
    }
}

File: app/src/main/java/com/shevelev/comics_viewer/common/threads/CancelationToken.kt
package com.shevelev.comics_viewer.common.threads

import java.util.concurrent.locks.ReentrantLock

class CancelationToken : ICancelationTokenRead {
    private var isCanceled = false
    private val lock = ReentrantLock()

    override fun isCanceled(): Boolean {
        var isCanceled = false
        lock.lock()
        isCanceled = try {
            this.isCanceled
        } finally {
            lock.unlock()
        }
        return isCanceled
    }

    fun cancel() {
        lock.lock()
        try {
            isCanceled = true
        } finally {
            lock.unlock()
        }
    }
}